<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Module: fio</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Module: fio</h1>

    




<section>

<header>
    
        
            
        
    
</header>

<article>
    <div class="container-overview">
    
        
            <div class="description">Methods for loading score statistics and covariance matrices from local files. This is primarily used by the CLI.</div>
        

        
            















<dl class="details">

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-license">License:</dt>
    <dd class="tag-license"><ul class="dummy"><li>MIT</li></ul></dd>
    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line1">line 1</a>
    </li></ul></dd>
    

    

    

    
</dl>


















        
    
    </div>

    

    

    
        <h3 class="subsection-title">Classes</h3>

        <dl>
            <dt><a href="module-fio-GenotypeCovarianceMatrix.html">GenotypeCovarianceMatrix</a></dt>
            <dd></dd>
        
            <dt><a href="module-fio-ScoreStatTable.html">ScoreStatTable</a></dt>
            <dd></dd>
        
            <dt><a href="module-fio-VariantMask.html">VariantMask</a></dt>
            <dd></dd>
        </dl>
    

    

    

    

    
        <h3 class="subsection-title">Members</h3>

        
            
<h4 class="name" id="~STATS_FORMAT"><span class="type-signature">(inner, constant) </span>STATS_FORMAT<span class="type-signature"> :Object</span></h4>




<div class="description">
    An enum denoting score/covariance statistics file format.
</div>



    <h5>Type:</h5>
    <ul>
        <li>
            
<span class="param-type">Object</span>


        </li>
    </ul>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line18">line 18</a>
    </li></ul></dd>
    

    

    

    
</dl>






        
    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id="~detectFormat"><span class="type-signature">(async, inner) </span>detectFormat<span class="signature">(fpath)</span><span class="type-signature"> &rarr; {number}</span></h4>
    

    



<div class="description">
    Determine whether the file is in rvtest or raremetal format.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>fpath</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">Path to file (can be covariance or score stats).</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line589">line 589</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Returns:</h5>

        
<div class="param-desc">
    STATS_FORMAT.RAREMETAL or STATS_FORMAT.RVTEST.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">number</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id="~extractCovariance"><span class="type-signature">(async, inner) </span>extractCovariance<span class="signature">(fpath, region, variants, scoreStats)</span><span class="type-signature"> &rarr; {Promise.&lt;GenotypeCovarianceMatrix>}</span></h4>
    

    



<div class="description">
    Extract covariance matrix from a file.
If variants are provided, only extract a matrix for the given variants. This only requires a single pass of the file.
If no variants are provided, a double pass is done - one to figure out the size of the matrix, the next to read it.

<p>

This function assumes you have tabix installed, and it exists on your PATH.
You can download tabix by visiting <a href="http://www.htslib.org/download/">htslib.org</a>.

<p>

We assume that the file being loaded is a covariance matrix file produced by either
<a href="https://genome.sph.umich.edu/wiki/RAREMETAL_Documentation">RAREMETAL</a> or
<a href="https://github.com/zhanxw/rvtests">rvtests</a>. We specifically assume that, per these file formats, scores and
covariances are oriented towards the alternate allele when reading. However, when storing, covariances are stored flipped
towards the *minor allele*, as this is typically the convention used in aggregation tests. The covariance matrix is
also multiplied by the sample size, since per convention, RAREMETAL and rvtests both divide each element by the sample
size before writing out.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>fpath</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">Path to covariance matrix file.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>region</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">Region string, e.g. 1:1-40000.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>variants</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array.&lt;string></span>


            
            </td>

            

            

            <td class="description last">Array of variants to extract in this order. Variants should be EPACTS format, e.g. 1:4_A/G.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>scoreStats</code></td>
            

            <td class="type">
            
                
<span class="param-type">ScoreStatTable</span>


            
            </td>

            

            

            <td class="description last">Object containing score statistics and other required information.
  This is needed because rvtest and raremetalworker both normalize the covariance matrix by the sample size.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line641">line 641</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Returns:</h5>

        
<div class="param-desc">
    A genotype covariance matrix.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Promise.&lt;GenotypeCovarianceMatrix></span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id="~extractScoreStats"><span class="type-signature">(async, inner) </span>extractScoreStats<span class="signature">(fpath, region, variants)</span><span class="type-signature"> &rarr; {module:stats~ScoreStatTable}</span></h4>
    

    



<div class="description">
    Extract score statistics from a file (either rvtest or raremetal format).
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>fpath</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">The path to the bgzipped score statistics file (one variant per line).</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>region</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">Region containing the variants. Should be formatted in the typical "1:1-4000".</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>variants</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array.&lt;string></span>


            
            </td>

            

            

            <td class="description last">A list of variants to specifically extract, in this order. If a list of variants is not
 provided, all variants will be extracted in the region.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line468">line 468</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Returns:</h5>

        
<div class="param-desc">
    An object containing statistics per variant, including:
 <ul>
   <li> Chromosome and position
   <li> Score statistic
   <li> Variance of the score statistic
   <li> Reference allele
   <li> Alternate allele and frequency
   <li> Effect allele and frequency
   <li> Number of genotyped samples present in the analysis when the score statistics were calculated
 </ul>
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">module:stats~ScoreStatTable</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id="~getNumberOfVariantsFromCovarianceFile"><span class="type-signature">(inner) </span>getNumberOfVariantsFromCovarianceFile<span class="signature">(covarFile, region)</span><span class="type-signature"> &rarr; {number}</span></h4>
    

    



<div class="description">
    Find the number of variants in a region of a covariance matrix file.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>covarFile</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">Path to covariance matrix file</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>region</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">Region string, e.g. 1:1-4000</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line570">line 570</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Returns:</h5>

        
<div class="param-desc">
    Number of variants in the region
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">number</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id="~readMaskFileSync"><span class="type-signature">(inner) </span>readMaskFileSync<span class="signature">(fpath)</span><span class="type-signature"> &rarr; {VariantMask}</span></h4>
    

    



<div class="description">
    Read groups from a mask file.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>fpath</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">Path to mask file.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line422">line 422</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Returns:</h5>

        
<div class="param-desc">
    A VariantMask object that stores a mapping of groups to lists of variants.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">VariantMask</span>


    </dd>
</dl>

    





        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Modules</h3><ul><li><a href="module-browser.html">browser</a></li><li><a href="module-cli.html">cli</a></li><li><a href="module-fio.html">fio</a></li><li><a href="module-print.html">print</a></li><li><a href="module-qfc.html">qfc</a></li><li><a href="module-rstats.html">rstats</a></li><li><a href="module-stats.html">stats</a></li></ul><h3>Classes</h3><ul><li><a href="module-fio-GenotypeCovarianceMatrix.html">GenotypeCovarianceMatrix</a></li><li><a href="module-fio-ScoreStatTable.html">ScoreStatTable</a></li><li><a href="module-fio-VariantMask.html">VariantMask</a></li><li><a href="module-stats-AggregationTest.html">AggregationTest</a></li><li><a href="module-stats-SkatTest.html">SkatTest</a></li><li><a href="module-stats-ZegginiBurdenTest.html">ZegginiBurdenTest</a></li><li><a href="PortalVariantsHelper.html">PortalVariantsHelper</a></li></ul><h3>Global</h3><ul><li><a href="global.html#AGGREGATION_TESTS">AGGREGATION_TESTS</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Tue Jun 26 2018 16:33:51 GMT-0400 (EDT)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>