<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: GenotypeCovarianceMatrix</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: GenotypeCovarianceMatrix</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>
            <span class="ancestors"><a href="module-fio.html">fio</a>~</span>GenotypeCovarianceMatrix<span class="signature">(matrix, variants, positions)</span><span class="type-signature"></span></h2>
        
            <div class="class-description">Class for storing genotype covariance matrices. <br/><br/>

Assumptions:
<ul>
 <li> Covariances should be oriented towards the minor allele.
 <li> Variances on the diagonal are absolute values (they are not directional.)
</ul></div>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    
    <h2>Constructor</h2>
    

    
    <h4 class="name" id="GenotypeCovarianceMatrix"><span class="type-signature"></span>new GenotypeCovarianceMatrix<span class="signature">(matrix, variants, positions)</span><span class="type-signature"></span></h4>
    

    











    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>matrix</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array.&lt;Array.&lt;number>></span>


            
            </td>

            

            

            <td class="description last">Pre-constructed matrix. Usually generated by the
 <a href="module-fio.html#~extractCovariance">extractCovariance</a> function in fio.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>variants</code></td>
            

            <td class="type">
            
                
<span class="param-type">Map</span>


            
            </td>

            

            

            <td class="description last">Map of variants -> matrix position. Variants should be chr:pos_ref/alt format.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>positions</code></td>
            

            <td class="type">
            
                
<span class="param-type">Map</span>


            
            </td>

            

            

            <td class="description last">Map of variant position -> matrix position. Both positions should be integers.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line314">line 314</a>
    </li></ul></dd>
    

    

    

    
</dl>




















    
    </div>

    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id="add"><span class="type-signature"></span>add<span class="signature">(other)</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    Combine this covariance matrix with another.
This operation happens in place; this matrix will be overwritten with the new one.
Used in meta-analysis, see <a href="https://genome.sph.umich.edu/wiki/RAREMETAL_METHOD#SINGLE_VARIANT_META_ANALYSIS">our wiki</a>
for more information.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>other</code></td>
            

            <td class="type">
            
                
<span class="param-type">GenotypeCovarianceMatrix</span>


            
            </td>

            

            

            <td class="description last">Another covariance matrix</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line367">line 367</a>
    </li></ul></dd>
    

    

    

    
</dl>




















        
            

    

    
    <h4 class="name" id="dim"><span class="type-signature"></span>dim<span class="signature">()</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    Return dimensions of matrix
</div>













<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line352">line 352</a>
    </li></ul></dd>
    

    

    

    
</dl>




















        
            

    

    
    <h4 class="name" id="isComplete"><span class="type-signature"></span>isComplete<span class="signature">()</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    Determine whether matrix is complete.
Can specify i, j which are actual indices, or positions pos_i, pos_j which represent the positions of the variants.
</div>













<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="fio.js.html">fio.js</a>, <a href="fio.js.html#line326">line 326</a>
    </li></ul></dd>
    

    

    

    
</dl>




















        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Modules</h3><ul><li><a href="module-browser.html">browser</a></li><li><a href="module-cli.html">cli</a></li><li><a href="module-fio.html">fio</a></li><li><a href="module-print.html">print</a></li><li><a href="module-qfc.html">qfc</a></li><li><a href="module-quadrature.html">quadrature</a></li><li><a href="module-rstats.html">rstats</a></li><li><a href="module-stats.html">stats</a></li></ul><h3>Classes</h3><ul><li><a href="module-fio-GenotypeCovarianceMatrix.html">GenotypeCovarianceMatrix</a></li><li><a href="module-fio-ScoreStatTable.html">ScoreStatTable</a></li><li><a href="module-fio-VariantMask.html">VariantMask</a></li><li><a href="module-quadrature-ExpSinh.html">ExpSinh</a></li><li><a href="module-stats-AggregationTest.html">AggregationTest</a></li><li><a href="module-stats-SkatOptimalTest.html">SkatOptimalTest</a></li><li><a href="module-stats-SkatTest.html">SkatTest</a></li><li><a href="module-stats-VTTest.html">VTTest</a></li><li><a href="module-stats-ZegginiBurdenTest.html">ZegginiBurdenTest</a></li><li><a href="PortalTestRunner.html">PortalTestRunner</a></li><li><a href="PortalVariantsHelper.html">PortalVariantsHelper</a></li></ul><h3>Global</h3><ul><li><a href="global.html#AGGREGATION_TESTS">AGGREGATION_TESTS</a></li><li><a href="global.html#cholesky">cholesky</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc/jsdoc">JSDoc 3.6.5</a> on Mon Nov 23 2020 16:22:01 GMT-0500 (Eastern Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>